src/intrinsics/prepack/global.js

```diff
@@ -400,4 +401,38 @@ export default function(realm: Realm): void {
     enumerable: false,
     configurable: true,
   });
+
+  global.$DefineOwnProperty("__safeSideEffect", {
+    value: new NativeFunctionValue(realm, "global.__safeSideEffect", "__safeSideEffect", 1, (context, [func]) => {
+      invariant(func instanceof ECMAScriptSourceFunctionValue || func instanceof BoundFunctionValue);
+      func.kind = "sideEffectful wrapper";
+      // TODO: We can properly replace this logic with some custom logic rather than
+      // using the havoc system at some point. Right now, we use the havoc system to
+      // find all the bindings/properties within the "safe side effect" closure. It
+      // marks them as havoced, which means their values gets the "havoced" flag applied.
+      // With this flag, any accesses to them results in abstract values being returned
+      // making it possible to mutate/reference values in the pure function.
+
+      // store the modifiedBindings before we havoc the function
+      let originalModifiedBindings = new Map(realm.modifiedBindings);
+      let originalModifiedProperties = new Map(realm.modifiedProperties);
+      // use the havoc system to mark all the bindings and modifications to objects
+      // in the function as havoced, thus making the values become abstract
+      if (func instanceof ECMAScriptSourceFunctionValue) {
+        Havoc.value(realm, func);
+      } else if (func instanceof BoundFunctionValue) {
+        Havoc.value(realm, func.$BoundTargetFunction);
+      }
+      // given we don't actually want the modified bindings/properties from the havoc
+      // as this is a side-effect and such, we can discard the modified bindings/properties
+      realm.modifiedBindings = originalModifiedBindings;
```

Comment:
I don't fully understand the intention here, but this looks wrong and dangerous to me. So `Havoc` changed some bindings, but removing the new entries from `realm.modifiedBindings` means that these changed bindings won't participate in proper tracking and undoing of effects later on.

Reply:
We could diff the changes to bindings and then ignore them specifically when checking for side-effects? I originally tried that but ran into some strange errors.

Reply:
Yea, this is way too fragile.

The indention here is that we want to havoc something outside the pure scope which we never do in any other case, and we want that havoc to be permanent so it shouldn't be reverted.

I'm not sure this is the best strategy though. Not sure havocing outside the pure function actually works as expected. Will need to think about that.